python_modules/dagster/dagster/_core/op_concurrency_limits_counter.py

```diff
@@ -0,0 +1,120 @@
+from collections import defaultdict
+from typing import Optional, Sequence
+
+import pendulum
+
+from dagster._core.snap.execution_plan_snapshot import ExecutionPlanSnapshot
+
+from .instance import DagsterInstance
+from .storage.dagster_run import DagsterRun, DagsterRunStatus, RunRecord
+from .storage.tags import GLOBAL_CONCURRENCY_TAG
+
+
+def compute_root_concurrency_keys_for_snapshot(
+    plan_snapshot: ExecutionPlanSnapshot,
+) -> Sequence[Optional[str]]:
+    """Utility function called at run creation time to add the concurrency info needed to keep track
+    of concurrency limits for each in-flight run.
+    """
+    root_step_keys = set(
+        [step_key for step_key, deps in plan_snapshot.step_deps.items() if not deps]
+    )
+    return [
+        (step.tags or {}).get(GLOBAL_CONCURRENCY_TAG)
+        for step in plan_snapshot.steps
+        if step.key in root_step_keys
+    ]
+
+
+class GlobalOpConcurrencyLimitsCounter:
+    def __init__(
+        self,
+        instance: DagsterInstance,
+        runs: Sequence[DagsterRun],
+        in_progress_run_records: Sequence[RunRecord],
+        slot_count_offset: int = 0,
+        started_run_buffer_seconds: int = 0,
+    ):
+        self._root_concurrency_keys_by_run = {}
+        self._concurrency_info_by_key = {}
+        self._launched_concurrency_key_counts = defaultdict(int)
+        self._in_progress_concurrency_key_counts = defaultdict(int)
+        self._slot_count_offset = slot_count_offset
+        self._started_run_buffer_seconds = started_run_buffer_seconds
+
+        # fetch all the concurrency info for all of the runs at once, so we can claim in the correct
+        # priority order
+        self._fetch_concurrency_info(instance, runs)
+
+        # fetch all the outstanding concurrency keys for in-progress runs
+        self._process_in_progress_runs(in_progress_run_records)
+
+    def _fetch_concurrency_info(self, instance: DagsterInstance, queued_runs: Sequence[DagsterRun]):
+        # fetch all the concurrency slot information for the root concurrency keys of all the queued
+        # runs
+        all_concurrency_keys = set()
+        for run in queued_runs:
+            if not run.root_op_concurrency_keys:
+                continue
+            all_concurrency_keys.update([key for key in run.root_op_concurrency_keys if key])
+
+        for key in all_concurrency_keys:
+            self._concurrency_info_by_key[key] = instance.event_log_storage.get_concurrency_info(
+                key
+            )
+
+    def _should_allocate_slots_for_root_concurrency_keys(self, record: RunRecord):
+        status = record.dagster_run.status
+        if status == DagsterRunStatus.STARTING:
+            return True
+        if status != DagsterRunStatus.STARTED or not record.start_time:
```

Comment:
start_time is always set exactly when the run moves into STARTED right?

Reply:
that's right, yes